packages/store/tests/action-handler.spec.ts

```diff
@@ -0,0 +1,227 @@
+      store.dispatch(new TestAction());
+      store.reset({ foo: newState });
+      // Act
+      tick();
+      // Assert
+      expect(currentState).toBe(newState);
+    }));
+
+    it(`should allow for the state to be set during the callback`, fakeAsync(() => {
+      // Arrange
+      @State({ name: 'foo', defaults: { name: 'old state' } })
+      class FooState {
+        @Action(TestAction)
+        test({ setState }: StateContext<any>, { payload }: TestAction) {
+          return timer(0).pipe(
+            tap(() => {
+              setState(payload);
+            })
+          );
+        }
+      }
+
+      const { store } = setup({ stores: [FooState] });
+      const newState = { name: 'new state' };
+      store.dispatch(new TestAction(newState));
+      const stateBeforeAsync = store.selectSnapshot(FooState);
+      // Act
+      tick();
+      // Assert
+      const resultState = store.selectSnapshot(FooState);
+      expect(resultState).toBe(newState);
+      expect(resultState).not.toEqual(stateBeforeAsync);
+    }));
+
+    it(`should allow for patching the state during the callback`, fakeAsync(() => {
+      // Arrange
+      @State({ name: 'foo', defaults: { name: 'my state' } })
+      class FooState {
+        @Action(TestAction)
+        test({ patchState }: StateContext<any>, { payload }: TestAction) {
+          return timer(0).pipe(
+            tap(() => {
+              patchState({ age: payload });
+            })
+          );
+        }
+      }
+
+      const { store } = setup({ stores: [FooState] });
+      store.dispatch(new TestAction(21));
+      const stateBeforeAsync = store.selectSnapshot(FooState);
+      // Act
+      tick();
+      // Assert
+      const resultState = store.selectSnapshot(FooState);
+      const expectedState = { name: 'my state', age: 21 };
+      expect(resultState).toEqual(expectedState);
+      expect(resultState).not.toEqual(stateBeforeAsync);
+    }));
+
+    it(`should allow for patching the state using a function during the callback`, fakeAsync(() => {
+      // Arrange
+      @State({ name: 'foo', defaults: { name: 'my state' } })
+      class FooState {
+        @Action(TestAction)
+        test({ patchState }: StateContext<any>, { payload }: TestAction) {
+          return timer(0).pipe(
+            tap(() => {
+              patchState(existing => {
+                return { ...existing, age: payload, updated: true };
+              });
+            })
+          );
+        }
+      }
+
+      const { store } = setup({ stores: [FooState] });
+      store.dispatch(new TestAction(21));
+      const stateBeforeAsync = store.selectSnapshot(FooState);
+      // Act
+      tick();
+      // Assert
+      const resultState = store.selectSnapshot(FooState);
+      const expectedState = { name: 'my state', age: 21, updated: true };
+      expect(resultState).toEqual(expectedState);
+      expect(resultState).not.toEqual(stateBeforeAsync);
```

Comment:
Where test with Immer?

Reply:
Not necessary because it was just used as an example in the docs.
I'm also not too keen to take a dependency on `immer` just for the tests.